### DISCUSSION TITLE
Normal stress-  error: 'declareProperty' was not declared in this scope

### DISCUSSION THREAD

**Discussion by Rahim-Habibi** [idaholab/moose#21945] [Q&A Modules: Solid mechanics] [answered]
Dear all, 

I was looking for a way to calculate "Normal Stress" to model Mohr-Coulomb Failure criterion. Hopefully, I found that "_normals" represents the normals of element face which is also used in "pressure BC" in Tensor Mechanics. So I developed the header and source file, but it does not compile, shows the error: 
```
In file included from /mnt/c/For_MOOSE/projects/tiger/build/unity_src/auxkernels_Unity.C:3:
/mnt/c/For_MOOSE/projects/tiger/src/auxkernels/TigerMohrCoulombFailureCriterion.C: In constructor 'TigerMohrCoulombFailureCriterion::TigerMohrCoulombFailureCriterion(const InputParameters&)':
/mnt/c/For_MOOSE/projects/tiger/src/auxkernels/TigerMohrCoulombFailureCriterion.C:36:18: error: 'declareProperty' was not declared in this scope
   36 |         _normals(declareProperty<RealVectorValue>("_normals")),
      |                  ^~~~~~~~~~~~~~~
/mnt/c/For_MOOSE/projects/tiger/src/auxkernels/TigerMohrCoulombFailureCriterion.C:36:49: error: expected primary-expression before '>' token
   36 |         _normals(declareProperty<RealVectorValue>("_normals")),
      |                                                 ^
make: *** [/mnt/c/For_MOOSE/projects/moose/framework/build.mk:145: /mnt/c/For_MOOSE/projects/tiger/build/unity_src/auxkernels_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1.
```

The .h :
```
#pragma once

#include "AuxKernel.h"
#include "MooseTypes.h"
#include "libmesh/quadrature_gauss.h"
#include "RankTwoTensor.h"

 class TigerMohrCoulombFailureCriterion : public AuxKernel
 {
 public:
   static InputParameters validParams();
   TigerMohrCoulombFailureCriterion(const InputParameters & parameters);

    protected:

    Real computeValue() override;

     unsigned int _component;
    const bool _use_displaced_mesh;
    const std::string _base_name;
    const MaterialProperty<RankTwoTensor> & _TenMech_total_stress;

    Moose::CoordinateSystemType _coord_type;
    const MaterialProperty<RealVectorValue>  & _normals;
    //const MooseArray<RealVectorValue> & _normals;
    MaterialProperty<Real>    _normals_stress;
 };
```
and .c files as follows:

```
#include "TigerMohrCoulombFailureCriterion.h"
#include "Material.h"
#include "Assembly.h"

registerMooseObject("TigerApp", TigerMohrCoulombFailureCriterion);

InputParameters

TigerMohrCoulombFailureCriterion::validParams()
{
  InputParameters params = AuxKernel::validParams();
  params.addClassDescription("calculates normal stress on a given boundary by normal vector");
  params.addDeprecatedParam<unsigned int>(
      "component", "The component for the pressure", "This parameter is no longer necessary");
  params.addParam<std::string>("base_name", "the identical base name provided "
        "in TensorMechanics Action");
  params.addRequiredParam<MaterialPropertyName>("total_stress",
                                                "The rank two material tensor name");

  params.set<bool>("use_displaced_mesh") = true;
  return params;
};
TigerMohrCoulombFailureCriterion::TigerMohrCoulombFailureCriterion(const InputParameters & parameters)
  : AuxKernel(parameters),
        _component(libMesh::invalid_uint),
        _use_displaced_mesh(getParam<bool>("use_displaced_mesh")),
        //_normals(getMaterialProperty<RealVectorValue>("normals")),
        _normals(declareProperty<RealVectorValue>("_normals")),
        //_normals(_assembly.adNormals<RealVectorValue>("_normals")),
        _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
        _TenMech_total_stress(getMaterialProperty<RankTwoTensor>(_base_name + "total_stress"))
{
}
  Real
  TigerMohrCoulombFailureCriterion::computeValue()
 {
       _normals_stress[_qp] = (_normals[_qp] * (_TenMech_total_stress[_qp])) * (_normals[_qp]);

        return 0;
 }
```
Thanks in advance to any suggestion to solve this error. 

Rahim

**Comment by GiudGiud** [comment, marked as answer]
Hello

declareProperty is for material properties, defined (and declared) in Materials.
You are writing an auxkernel here. 
Maybe you meant to retrieve a material property instead? With getMaterialProperty

Guillaume

**Reply by Rahim-Habibi** [reply]
Hi,
Sorry for late back, actually, my purpose is to use normal vector of element face by applying "_normals". Already, the normal vector was used in "PressureBC" and [here](https://github.com/ngrilli/c_pfor_am/blob/main/src/dgkernels/DGAdvectionCoupled.C) by @ngrilli. 
So, based on these and your suggestion, I editted my Auxkernel:
1- I used " const MooseArray<Point> & _normals; " instead of "const MaterialProperty<RealVectorValue>  & _normals;" in .h file.
2- and, " _normals(_assembly.normals("_normals"))," instead of declarying "_normals(declareProperty<RealVectorValue>("_normals")),". 
Now, It is linked to my App, but once I am running a simple example,  it shows following error:
```
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 27330 RUNNING AT
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

BTW, I used it in my input file like:
[./direction2]
    type = TigerMohrCoulombFailureCriterion
    total_stress = stress
    variable = direction2
    execute_on = 'TIMESTEP_END'
    block = 'Fault'
  [../]

**Reply by GiudGiud** [reply]
Hello

could you please follow the debugging instructions on the website and get us a backtrace? It should tell us which line you are getting a segfault on

Guillaume

**Reply by Rahim-Habibi** [reply]
Hi, 
I debuged based on the instruction on that page: first build a debug executable then I debuged using "gdb". The following shows the results. BTW I tried to run after debuging, but still shows the same error. 
```
(moose) rahim-agw@AGW-GEOT-STUD15:/mnt/c/For_MOOSE/projects/tiger$ gdb --args ./tiger-dbg
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04.1) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./tiger-dbg...
(gdb) run
Starting program: /mnt/c/For_MOOSE/projects/tiger/tiger-dbg
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffe85f0700 (LWP 7558)]

Usage: /mnt/c/For_MOOSE/projects/tiger/tiger-dbg [<options>]

Options:
  --allow-test-objects                              Register test objects and syntax.
  --app <AppName>                                   Specify the application that should be used to run the input file. This must match an application name registered to the application factory. Note that this option is case-sensitive.
  --type                                            Return the name of the application object.
  --check-input                                     Check the input file (i.e. requires -i <filename>) and quit.
  --color [auto,on,off]                             Whether to use color in console output (default 'on').
  --copy-tests                                      copy installed tests to an [appname]_tests dir
  --definition                                      Shows a SON style input definition dump for input validation
  --disable-perf-graph-live                         Disables PerfGraph Live Printing.
  -v --version                                      Print application version
  --distributed-mesh                                The libMesh Mesh underlying MooseMesh should always be a DistributedMesh
  --dump [search_string]                            Shows a dump of available input file syntax.
  --error                                           Turn all warnings into errors
  --error-deprecated                                Turn deprecated code messages into Errors
  -o --error-override                               Error when encountering overridden or parameters supplied multiple times
  -e --error-unused                                 Error when encountering unused input file options
  --half-transient                                  When true the simulation will only run half of its specified transient (ie half the timesteps).  This is useful for testing recovery and restart
  -h --help                                         Displays CLI usage statement.
  -i <input_files>                                  Specify one or multiple input files. Multiple files get merged into a single simulation input.
  --json                                            Dumps input file syntax in JSON format.
  --keep-cout                                       Keep standard output from all processors when running in parallel
  --list-constructed-objects                        List all moose object type names constructed by the master app factory.
  --mesh-only [mesh_file_name]                      Setup and Output the input mesh only (Default: "<input_file_name>_in.e")
  --minimal                                         Ignore input file and build a minimal application with Transient executioner.
  --n-threads=<n>                                   Runs the specified number of threads per process
  --no-color                                        Disable coloring of all Console outputs.
  --no-timing                                       Disabled performance logging. Overrides -t or --timing if passed in conjunction with this flag
  --no-trap-fpe                                     Disable Floating Point Exception handling in critical sections of code when using DEBUG mode.
  --perf-graph-live-all                             Forces printing of ALL progress messages.
  --recover [file_base]                             Continue the calculation.  If file_base is omitted then the most recent recovery file will be utilized
  --recoversuffix [suffix]                          Use a different file extension, other than cpr, for a recovery file
  --redirect-stdout                                 Keep standard output from all processors when running in parallel
  -r <n>                                            Specify additional initial uniform refinements for automatic scaling
  --registry                                        Lists all known objects and actions.
  --registry-hit                                    Lists all known objects and actions in hit format.
  --tests                                           run all tests
  --show-controls                                   Shows the Control logic available and executed.
  --docs                                            print url/path to the documentation website
  --show-input                                      Shows the parsed input file before running the simulation.
  --show-outputs                                    Shows the output execution time information.
  --split-file [filename]                           optional name of split mesh file(s) to write/read
  --split-mesh [splits]                             comma-separated list of numbers of chunks to split the mesh into
  --start-in-debugger <debugger>                    Start the application and attach a debugger.  This will launch xterm windows using the command you specify for 'debugger'
  --stop-for-debugger [seconds]                     Pauses the application during startup for the specified time to allow for connection of debuggers.
  --syntax                                          Dumps the associated Action syntax paths ONLY
  -t --timing                                       Enable all performance logging for timing purposes. This will disable all screen output of performance logs for all Console objects.
  --trap-fpe                                        Enable Floating Point Exception handling in critical sections of code.  This is enabled automatically in DEBUG mode
  --executor                                        Use the new Executor system instead of Executioners
  --use-split                                       use split distributed mesh files
  -w --warn-unused                                  Warn about unused input file options
  --yaml                                            Dumps input file syntax in YAML format.

Solver Options:
  See solver manual for details (Petsc or Trilinos)
[Thread 0x7fffe85f0700 (LWP 7558) exited]

 ----------------------------------------------------------------------------
| Reference count information                                                |
 ----------------------------------------------------------------------------
| N7libMesh10Parameters5ValueE reference count information:
|  Creations:    311
|  Destructions: 311
| N7libMesh4ElemE reference count information:
|  Creations:    1
|  Destructions: 1
| N7libMesh9DofObjectE reference count information:
|  Creations:    1
|  Destructions: 1
 ----------------------------------------------------------------------------
[Inferior 1 (process 7554) exited normally]
(gdb)
```

**Reply by GiudGiud** [reply]
you need to run with the input file to get the backtrace
`run -i <input_file.i>`

**Reply by GiudGiud** [reply]
I think _normals should be initiliazed like this:
``
_normals(_assembly.normals()),
``
with no string in the function call like you pasted?

**Reply by Rahim-Habibi** [reply]
Hi, 

I debuged the input file, but it shows following error: 

```
Assertion `i < _size' failed
Access out of bounds in MooseArray (i: 0 size: 0)
at /mnt/c/For_MOOSE/projects/moose/framework/build/header_symlinks/MooseArray.h, line 276
Stack frames: 22
0: libMesh::print_trace(std::ostream&)
1: /mnt/c/For_MOOSE/projects/tiger/lib/libtiger-dbg.so.0(+0x14ac2d) [0x7fffff65dc2d]
2: TigerMohrCoulombFailureCriterion::computeValue()
3: AuxKernelTempl<double>::compute()
4: ComputeElemAuxVarsThread<AuxKernelTempl<double> >::onElement(libMesh::Elem const*)
5: ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
6: /mnt/c/For_MOOSE/projects/moose/framework/libmoose-dbg.so.0(+0x34bd1c0) [0x7ffff58441c0]
7: void AuxiliarySystem::computeElementalVarsHelper<AuxKernelTempl<double> >(MooseObjectWarehouse<AuxKernelTempl<double> > const&, std::vector<std::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> >, std::allocator<std::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> > > > const&)
8: AuxiliarySystem::computeElementalVars(MooseEnumItem)
9: AuxiliarySystem::compute(MooseEnumItem)
10: FEProblemBase::computeAuxiliaryKernels(MooseEnumItem const&)
11: FEProblemBase::execute(MooseEnumItem const&)
12: FixedPointSolve::solveStep(double&, double&, std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&)
13: FixedPointSolve::solve()
14: TimeStepper::step()
15: Transient::takeStep(double)
16: Transient::execute()
17: MooseApp::executeExecutioner()
18: MooseApp::run()
19: /mnt/c/For_MOOSE/projects/tiger/tiger-dbg(+0xa7fb) [0x800a7fb]
20: __libc_start_main
21: /mnt/c/For_MOOSE/projects/tiger/tiger-dbg(+0xa089) [0x800a089]
[0] /mnt/c/For_MOOSE/projects/moose/framework/build/header_symlinks/MooseArray.h, line 276, compiled Aug 29 2022 at 14:39:18
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
[Thread 0x7fffe90778c0 (LWP 512) exited]
[Inferior 1 (process 512) exited with code 01]
```

BTW, I went through to discussion forum, I found #21920 started almost a week ago but not answered so far. 
Regarding to your suggestion on how to initialize normal vector, I could say I initialized it already as you recommended.

**Reply by GiudGiud** [reply]
somewhere in 2: TigerMohrCoulombFailureCriterion::computeValue()
 you are accessing an array out of bounds. Could you run this in a debugger, add a breakpoint on MPI Abort and get us the exact line

**Reply by Rahim-Habibi** [reply]
I added breakpoint at that line, the following shows the backtrace:

Thread 1 "tiger-dbg" hit Breakpoint 1, 0x00007fffea084880 in PMPI_Abort ()
   from /home/rahim-agw/miniconda3/envs/moose/lib/libmpi.so.12
(gdb) bt
#0  0x00007fffea084880 in PMPI_Abort () from /home/rahim-agw/miniconda3/envs/moose/lib/libmpi.so.12
#1  0x00007fffff65dc6b in MooseArray<libMesh::Point>::operator[] (this=0x8938990, i=0)
    at /mnt/c/For_MOOSE/projects/moose/framework/build/header_symlinks/MooseArray.h:276
#2  0x00007fffff6591b3 in TigerMohrCoulombFailureCriterion::computeValue (this=0x8cccd60)
    at /mnt/c/For_MOOSE/projects/tiger/src/auxkernels/TigerMohrCoulombFailureCriterion.C:47
#3  0x00007ffff48ca2a2 in AuxKernelTempl<double>::compute (this=0x8cccd60)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/auxkernels/AuxKernel.C:295
#4  0x00007ffff4e624f8 in ComputeElemAuxVarsThread<AuxKernelTempl<double> >::onElement (this=0x7ffffffeb580,
    elem=0x888dcd0) at /mnt/c/For_MOOSE/projects/moose/framework/src/loops/ComputeElemAuxVarsThread.C:115
#5  0x00007ffffb9e357f in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator() (this=0x7ffffffeb580, range=..., bypass_threading=false)
    at /mnt/c/For_MOOSE/projects/moose/framework/build/header_symlinks/ThreadedElementLoopBase.h:226
#6  0x00007ffff58441c0 in libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeElemAuxVarsThread<AuxKernelTempl<double> > > (range=..., body=...)
    at /home/rahim-agw/miniconda3/envs/moose/libmesh/include/libmesh/threads_pthread.h:380
#7  0x00007ffff57f0919 in AuxiliarySystem::computeElementalVarsHelper<AuxKernelTempl<double> > (this=0x88df290,
    warehouse=..., vars=...) at /mnt/c/For_MOOSE/projects/moose/framework/src/systems/AuxiliarySystem.C:830
#8  0x00007ffff57eff61 in AuxiliarySystem::computeElementalVars (this=0x88df290, type=...)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/systems/AuxiliarySystem.C:756
#9  0x00007ffff57ecacc in AuxiliarySystem::compute (this=0x88df290, type=...)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/systems/AuxiliarySystem.C:445
#10 0x00007ffff55869e7 in FEProblemBase::computeAuxiliaryKernels (this=0x88bc8f0, type=...)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/problems/FEProblemBase.C:3737
#11 0x00007ffff55868e4 in FEProblemBase::execute (this=0x88bc8f0, exec_type=...)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/problems/FEProblemBase.C:3722
#12 0x00007ffff4bb7b9d in FixedPointSolve::solveStep (this=0x8917510, begin_norm=@0x9288530: 0,
    end_norm=@0x924d120: 0, transformed_dofs=...)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/executioners/FixedPointSolve.C:426
#13 0x00007ffff4bb6b7a in FixedPointSolve::solve (this=0x8917510)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/executioners/FixedPointSolve.C:273
#14 0x00007ffff598cd24 in TimeStepper::step (this=0x8b580d0)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/timesteppers/TimeStepper.C:161
#15 0x00007ffff4bc631a in Transient::takeStep (this=0x8916ca0, input_dt=-1)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/executioners/Transient.C:444
#16 0x00007ffff4bc5640 in Transient::execute (this=0x8916ca0)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/executioners/Transient.C:329
#17 0x00007ffff5c4c533 in MooseApp::executeExecutioner (this=0x81ed6c0)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/base/MooseApp.C:1081
#18 0x00007ffff5c4fa4f in MooseApp::run (this=0x81ed6c0)
    at /mnt/c/For_MOOSE/projects/moose/framework/src/base/MooseApp.C:1465
#19 0x000000000800a7fb in main (argc=3, argv=0x7ffffffec7c8) at /mnt/c/For_MOOSE/projects/tiger/src/main.C:24

BTW, I am using MOOSE on UBUNTU APP on Windows.

**Reply by GiudGiud** [reply]
Ok so what is line 47 in TigerMohrCoulombFailure?
Would it be possible to give me access to tiger?

**Reply by Rahim-Habibi** [reply]
The line 47:  return MetaPhysicL::raw_value (_normals[_qp] * (_TenMech_total_stress[_qp]) * (_normals[_qp])); 
where _normals refers to normal vector on element face and _tenMech_total_stress gets stress state in current time step, which calculates normal stress. I other words, it returns the multiplication: vector * matrix* vector. 
Also, you can see .c file on top of the discussion. 
Regarding your access to tiger, I should say that tiger is already commit [https://github.com/MGK-Lab/tiger](url), but failure criterion is being developed by this auxkernel.

**Reply by GiudGiud** [reply]
ok so either the access into _normals or into _TenMech_total_stress at index _qp is wrong.
I d do it by elimination, or you can use a debugger, place a breakpoint in the function, then print the objects
I suspect the normals still

the .C file at the top of the discussion is outdated. You declare _normals as a material property. You should have changed it to:
    _normals(_assembly.normals())

Indeed I cant see the new file on the tiger repository

**Reply by Rahim-Habibi** [reply]
I put a breakpoint in the function and debugged it by the way given  [https://mooseframework.inl.gov/application_development/debugging.html](url), but I got the same back trace as above. 
I declared normal vector in header file like: const MooseArray <Point> & _normals;
and in source file : _normals(_assembly.normals()),
We already used "_TenMech_total_stress" in our app, so it seems that the declaration of normal vector is may be wrong. 
I really appreciate in advance to help me solve the issue

**Reply by GiudGiud** [reply]
normals should be 
```
  /// normals at quadrature points
  const MooseArray<Point> & _normals;
```

**Reply by Rahim-Habibi** [reply]
It is exactly the same I defined in header file. This part "<Point>" unfortunately was not shown in my previous reply. But I defined the normal vector as you mentioned.

**Reply by GiudGiud** [reply]
ah ok. Not sure then. 
Can you print it in the debugger?
as in print the local variables

**Reply by Rahim-Habibi** [reply]
I printed _normals in debugger:

(gdb) p _normals
$1 = (const MooseArray<libMesh::Point> &) @0x8938990: {_data_ptr = {_M_t = {
      _M_t = {<std::_Tuple_impl<0, libMesh::Point*, std::default_delete<libMesh::Point []> >> = {<std::_Tuple_impl<1, std::default_delete<libMesh::Point []> >> = {<std::_Head_base<1, std::default_delete<libMesh::Point []>, true>> = {<std::default_delete<libMesh::Point []>> = {<No data fields>}, <No data fields>}, <No data fields>}, <std::_Head_base<0, libMesh::Point*, false>> = {_M_head_impl = 0x0}, <No data fields>}, <No data fields>}}}, _data = 0x0, _size = 0,
  _allocated_size = 0}

**Reply by GiudGiud** [reply]
ok well it's confirmed! normals isnt initialized well. 

is this being used inside the domain or on a boundary

**Reply by Rahim-Habibi** [reply]
Inside the domain.

**Reply by GiudGiud** [reply]
so you know the integral of this quantity over the whole volume analytically?

**Reply by Rahim-Habibi** [reply]
let me describe it. Actually I wish to have normal vector on fault elements (2D element) within a geological model (box-shaped model). So, the elements I am interested in to have the normal, do not have volume while they are inside the domain. 

BTW, I tried to print a word in function "computeValue" like: 
```
Real
 TigerMohrCoulombFailureCriterion::computeValue()
 {
     std::cout << "test" << std::endl;
//   return MetaPhysicL::raw_value (_normals[_qp] * (_TenMech_total_stress[_qp]) * (_normals[_qp]));

   return 0;
 }
```
but, it is not  printed during run, in other words, it seems that function "computeValue" is being ignored during run.

**Reply by GiudGiud** [reply]
are you running in parallel?
if the elements are not in the root process' domain then the console is automatically silenced

could also be that auxkernels are not executed on lower dimensional elements.
@YaqiWang do you know if they are?

**Reply by Rahim-Habibi** [reply]
Yes, I am running in parallel.  
That is good point that the elements should be in root process domain.
Still,  I am looking forward to hearing more from @YagiWang.

**Reply by GiudGiud** [reply]
Then it's that. 
You need to pass `--keep-cout` to have output from all ranks in parallel
